docs/how-to/write-tests.md

```diff
@@ -6,3 +6,130 @@ Testing is essential to maintain the integrity and reliability of the codebase.
+@pytest.fixture
+async def sim_my_device(RE: RunEngine) -> MyDevice:
+    async with init_devices(mock=True):
+        sim_my_device = MyDevice("TEST:")
+    return sim_my_device
+
+
+def test_my_device_stage(sim_my_device: MyDevice, RE: RunEngine) -> None:
+    RE(bps.stage(sim_my_device, wait=True), wait=True)
+    get_mock_put(sim_my_device.signal_b).assert_called_once_with(OnOff.ON, wait=True)
+
+
+def test_my_device_unstage(sim_my_device: MyDevice, RE: RunEngine) -> None:
+    RE(bps.unstage(sim_my_device, wait=True), wait=True)
+    get_mock_put(sim_my_device.signal_b).assert_called_once_with(OnOff.OFF, wait=True)
+
+
+async def test_my_device_read(sim_my_device: MyDevice, RE: RunEngine) -> None:
+    prefix = sim_my_device.name
+    await assert_reading(
+        sim_my_device,
+        {
+            f"{prefix}-signal_a": partial_reading(ANY),
+            f"{prefix}-signal_b": partial_reading(ANY),
+        },
+    )
+
+```
+
+For testing the read output of a device, there are some handy functions defined in `ophyd-async` to aid with this. `assert_reading` allows us to compare the readings generated from a `Readable` device to the expected results. `partial_reading` wraps the value given in a mapping like `{"value": ANY}`, so we are actually checking that the reading matches the expected structure.
+
+```Python
+    prefix = sim_my_device.name
+    await assert_reading(
+        sim_my_device,
+        {
+            f"{prefix}-signal_a" : {"value": ANY},
+            f"{prefix}-signal_b" : {"value": ANY},
+        }
+    )
+```
+
+Once we have written the tests, we need to check that they pass and how much coverage there is. `pytest` should automatically detect your tests and display one of the following icons within the VSCode IDE:
+- Grey circle (not run yet)
+- Green circle with a tick (test passed)
+- Red circle with a cross (test failed)
+
+You can also check code coverage by right-clicking on the test icon and selecting `Run with coverage`. This will re-run the test and highlight line numbers in your file:
+- Green = covered by tests
+- Red = not covered (and ideally should be tested)
+
+To view all tests inside dodal, use the `Test Explorer` panel (flask icon) in VSCode. It displays all of your tests in a hierarchy, which you can run individually or in groups. It also shows the pass/fail icons mentioned above. Clicking on an item will run all tests beneath it in the hierarchy.
+
+You can also run all tests in the command line via the command:
+`tox -e tests`
+
+This will also display each Python module’s percentage of code coverage achieved.
+
+If you find that your tests are being skipped or not recognised by `pytest`, check for any syntax errors as this will block the tests being found. You can also check to see if there is an error output at the very bottom of the `Test Explorer` panel. This is usually caused by invalid syntax in your test file, or by circular dependencies in the code you are testing.
```

Comment:
I think most of this is IDE specific and could be reduced or cut.

It should at least defer to these already existing docs and if required improve them instead
https://diamondlightsource.github.io/dodal/main/how-to/coverage.html
https://diamondlightsource.github.io/dodal/main/how-to/run-tests.html